content/academy/puppeteer_playwright/page.md

```diff
@@ -0,0 +1,311 @@
+
+await sleep(10);
+await browser.close();
+</marked-tab>
+<marked-tab header="Puppeteeer" lang="javascript">
+import puppeteer from 'puppeteer';
+
+const sleep = (secs) => new Promise((r) => setTimeout(r, secs * 1000));
+
+const browser = await puppeteer.launch({ headless: false });
+
+const page = await browser.newPage();
+
+await page.goto('https://google.com/');
+
+// Click the "I understand" button
+await page.click('button + button:nth-child(2)');
+
+// Type the query into the search box
+await page.type('input[title]', 'hello world');
+
+// Press enter
+await page.keyboard.press('Enter');
+
+await sleep(10);
+await browser.close();
+</marked-tab>
+```
+
+When we run it, we leave off on the results page:
+
+![Google results page reached by headless browser]({{@asset puppeteer_playwright/images/google-results.webp}})
+
+### Waiting for elements/events
+
+In a perfect world, every piece of content served on a website would be loaded instantly. We don't live in a perfect world though, and often times it can take anywhere 1/10th of a second to a few seconds to load some content onto a page. Puppeteer and Playwright don't sit around waiting for the pages to load though - if we tell it to do something with an element that hasn't rendered yet, it'll start trying to do it (which will result in nasty errors). We've got to tell it to wait.
+
+Different events can be waited for using the various `waitFor...` methods offered, while elements can be waited for with `page.waitForSelector()` in both Playwright and Puppeteer. In Playwright you can also create a locator for an element and use the `locator.waitFor()` function.
+
+```marked-tabs
+<marked-tab header="Playwright" lang="javascript">
+// Create a pointer for the first result element
+const firstResult = page.locator(':nth-match(.g a, 1)');
+
+// Wait for the first result element to be visible
+await firstResult.waitFor();
+
+// or just
+// await page.waitForSelector('.g a');
+</marked-tab>
+<marked-tab header="Puppeteeer" lang="javascript">
+// Wait for the first result element to be visible
+await page.waitForSelector('.g a');
+</marked-tab>
+```
+
+### Getting general page data
+
+There are two main page functions which will return general data:
+
+1. `page.content()` will return the entire HTML content of the page.
+2. `page.title()` will return the title of the current page.
+
+For our case, we'll click on the first result, then utilize `page.title()` to grab the title and log it to the console:
+
+```marked-tabs
+<marked-tab header="Playwright" lang="javascript">
+// Click the first result
+await firstResult.click();
+
+// Grab hold of the page's title
+const title = await page.title();
+
+// Log the title to the console
+console.log(title);
+</marked-tab>
+<marked-tab header="Puppeteeer" lang="javascript">
+// Click the first result
+await page.click('.g a');
+
+// Grab hold of the page's title
+const title = await page.title();
+
+// Log the title to the console
+console.log(title);
```

Comment:
You're not waiting for the navigation to complete here. Not sure if playwright has some magic to do this, but in Puppeteer you always needed to do:


```js
await Promise.all([
  page.waitForNavigation(),
  page.click('button')
])

await page.title()
```

Reply:
@mnmkng In the previous section we're waiting for the button's selector (instead of waiting for navigation)

Reply:
The Puppeteer code works as is